usage/src/org/apache/cloudstack/quota/QuotaAlertManagerImpl.java

```diff
@@ -0,0 +1,471 @@
+            // + accountDomain.getUuid());
+
+            final StrSubstitutor templateEngine = new StrSubstitutor(optionMap);
+            final String subject = templateEngine.replace(emailTemplate.getTemplateSubject());
+            final String body = templateEngine.replace(emailTemplate.getTemplateBody());
+            try {
+                _emailQuotaAlert.sendQuotaAlert(emailRecipients, subject, body);
+                emailToBeSent.sentSuccessfully();
+            } catch (Exception e) {
+                s_logger.error(String.format("Unable to send quota alert email (subject=%s; body=%s) to account %s (%s) recipients (%s) due to error (%s)", subject, body, account.getAccountName(),
+                        account.getUuid(), emailRecipients, e));
+            }
+        } else {
+            s_logger.error(String.format("No quota email template found for type %s, cannot send quota alert email to account %s(%s)", emailType, account.getAccountName(), account.getUuid()));
+        }
+    }
+
+    class DeferredQuotaEmail {
+        AccountVO account;
+        QuotaAccountVO quotaAccount;
+        QuotaConfig.QuotaEmailTemplateTypes emailTemplateType;
+        BigDecimal quotaUsage;
+
+        public DeferredQuotaEmail(AccountVO account, QuotaAccountVO quotaAccount, BigDecimal quotaUsage, QuotaConfig.QuotaEmailTemplateTypes emailTemplateType) {
+            this.account = account;
+            this.quotaAccount = quotaAccount;
+            this.emailTemplateType = emailTemplateType;
+            this.quotaUsage = quotaUsage;
+        }
+
+        public DeferredQuotaEmail(AccountVO account, QuotaAccountVO quotaAccount, QuotaConfig.QuotaEmailTemplateTypes emailTemplateType) {
+            this.account = account;
+            this.quotaAccount = quotaAccount;
+            this.emailTemplateType = emailTemplateType;
+            this.quotaUsage = new BigDecimal(-1);
+        }
+
+        public AccountVO getAccount() {
+            return account;
+        }
+
+        public BigDecimal getQuotaBalance() {
+            return quotaAccount.getQuotaBalance();
+        }
+
+        public BigDecimal getQuotaUsage() {
+            return quotaUsage;
+        }
+
+        public QuotaConfig.QuotaEmailTemplateTypes getEmailTemplateType() {
+            return emailTemplateType;
+        }
+
+        public void sentSuccessfully() {
+            if (emailTemplateType == QuotaEmailTemplateTypes.QUOTA_STATEMENT) {
+                quotaAccount.setLastStatementDate(new Date());
+                _quotaAcc.update(quotaAccount.getAccountId(), quotaAccount);
+            }
+            else {
+                quotaAccount.setQuotaAlertDate(new Date());
+                quotaAccount.setQuotaAlertType(emailTemplateType.ordinal());
+                _quotaAcc.update(quotaAccount.getAccountId(), quotaAccount);
+            }
```

Comment:
These two ``if`` blocks only differ in that the ``else`` sets the alert type.  Consider simplifying the method in a manner such as:
```
public void sentSuccessfully() {

    quotaAccount.setLastStatementDate(new Date());
    _quotaAcc.update(quotaAccount.getAccountId(), quotaAccount);

    if (emailTemplateType != QuotaEmailTemplateTypes.QUOTA_STATEMENT) {                 
        quotaAccount.setQuotaAlertType(emailTemplateType.ordinal());
    }

}
```